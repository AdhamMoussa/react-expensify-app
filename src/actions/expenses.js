import database from "../firebase/firebase";
import moment from "moment";

export const addExpense = expense => {
  return {
    type: "ADD_EXPENSE",
    newExpense: {
      ...expense
    }
  };
};

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      amount = 0,
      note = "",
      createdAt = 0
    } = expenseData;
    const expense = { description, amount, note, createdAt };
    return database
      .ref("expenses")
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

export const editExpense = (id, updatedData) => {
  return {
    type: "EDIT_EXPENSE",
    id,
    updatedData
  };
};

export const startEditExpense = (id, updatedData) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .update({ ...updatedData })
      .then(() => {
        dispatch(
          editExpense(id, {
            ...updatedData,
            createdAt: moment(updatedData.createdAt)
          })
        );
      });
  };
};

export const removeExpense = id => {
  return {
    type: "REMOVE_EXPENSE",
    id
  };
};

export const startRemoveExpense = id => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense(id));
      });
  };
};

export const setExpenses = expenses => {
  return {
    type: "SET_EXPENSES",
    expenses
  };
};

export const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref("expenses")
      .once("value")
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(child => {
          expenses.push({
            id: child.key,
            ...child.val(),
            createdAt: moment(child.val().createdAt)
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
