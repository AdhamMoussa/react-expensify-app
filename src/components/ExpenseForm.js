import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends React.Component {
  state = {
    description: this.props.editedExpense
      ? this.props.editedExpense.description
      : "",
    amount: this.props.editedExpense ? this.props.editedExpense.amount : "",
    note: this.props.editedExpense ? this.props.editedExpense.note : "",
    createdAt: this.props.editedExpense
      ? this.props.editedExpense.createdAt
      : moment(),
    datePickerFocused: false,
    descriptionError: false,
    amountError: false
  };
  handleDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  handleAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
      this.setState(() => ({ amount }));
  };
  handleNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  handleFormSubmit = e => {
    e.preventDefault();
    // handle description error
    if (!this.state.description) {
      this.setState(() => ({
        descriptionError: true
      }));
    } else if (this.state.description && this.state.descriptionError) {
      this.setState(() => ({ descriptionError: false }));
    }
    // handle amount error
    if (!this.state.amount) {
      this.setState(() => ({
        amountError: true
      }));
    } else if (this.state.amount && this.state.amountError) {
      this.setState(() => ({ amountError: false }));
    }
    // Submit
    if (this.state.description && this.state.amount) {
      this.setState(() => ({
        descriptionError: false,
        amountError: false
      }));
      this.props.onSubmitHandler({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        <form className="form" onSubmit={this.handleFormSubmit}>
          <div className="form__group">
            <input
              id="description-input"
              type="text"
              name="description"
              value={this.state.description}
              onChange={this.handleDescriptionChange}
              placeholder="Expense Description"
              className="form__input"
              autoFocus
            />
            <p
              className={
                this.state.descriptionError
                  ? "error-msg error-msg--visible"
                  : "error-msg"
              }
            >
              Please insert a valid description of your expense
            </p>
          </div>
          <div className="form__group">
            <input
              id="amount-input"
              type="text"
              name="amount"
              value={this.state.amount}
              onChange={this.handleAmountChange}
              placeholder="Amount"
              className="form__input"
            />
            <p
              className={
                this.state.amountError
                  ? "error-msg error-msg--visible"
                  : "error-msg"
              }
            >
              Please insert a valid amount of your expense
            </p>
          </div>
          <div className="form__group">
            <SingleDatePicker
              date={this.state.createdAt}
              onDateChange={date => {
                if (date) {
                  this.setState({ createdAt: date });
                }
              }}
              focused={this.state.datePickerFocused}
              onFocusChange={({ focused }) =>
                this.setState({ datePickerFocused: focused })
              }
              numberOfMonths={1}
              isOutsideRange={day => false}
            />
          </div>
          <div className="form__group">
            <textarea
              id="note-input"
              name="note"
              value={this.state.note}
              onChange={this.handleNoteChange}
              placeholder="Expense Note"
              className="form__input"
            />
          </div>
          <div className="form__group">
            <button className="button" type="submit">
              Submit
            </button>
          </div>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
