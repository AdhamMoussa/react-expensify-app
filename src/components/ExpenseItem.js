import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseItem = props => {
  return (
    <tr className="expenses-list__item">
      <td>
        <Link className="expenses-list__item-description" to={`/edit/${props.id}`}>
          <h3>{props.description}</h3>
        </Link>
        <p className="expenses-list__item-date">{moment(props.createdAt).format("MMM Do, YYYY")}</p>
      </td>
      <td>
        <p className="expenses-list__item-amount">{numeral(props.amount).format("$0,0.00")}</p>
      </td>
    </tr>
  );
};

export default ExpenseItem;
