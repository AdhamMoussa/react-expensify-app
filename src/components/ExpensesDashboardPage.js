import React from 'react';
import ExpensesList from './ExpensesList';
import ExpensesListHeader from './ExpensesListHeader';

const ExpensesDashboardPage = (props) => {
  return (
    <div>
      <ExpensesListHeader />
      <ExpensesList />
    </div>
  )
};


export default ExpensesDashboardPage;
