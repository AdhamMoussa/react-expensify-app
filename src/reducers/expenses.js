const defaultExpensesState = [];

const expensesReducer = (state = defaultExpensesState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.newExpense];
    case "EDIT_EXPENSE":
      return state.map(
        expense =>
          expense.id === action.id
            ? { ...expense, ...action.updatedData }
            : expense
      );
    case "REMOVE_EXPENSE":
      return state.filter(expense => expense.id !== action.id);
    default:
      return state;
  }
};

export default expensesReducer;
