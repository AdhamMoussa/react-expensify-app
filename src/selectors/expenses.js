import moment from 'moment';

const getExpensesView = (expenses, filters) => {
  return expenses.filter((expense) => {
    const textMatch = expense.description.toLowerCase().includes(filters.text.toLowerCase());
    const startDateMatch = filters.startDate ? expense.createdAt.isSameOrAfter(filters.startDate) : true;
    const endDateMatch = filters.endDate ? expense.createdAt.isSameOrBefore(filters.endDate) : true;

    return textMatch && startDateMatch && endDateMatch;
  }).sort((a, b) => {
    if (filters.sortBy === 'date') {
      return a.createdAt.isAfter(b.createdAt) ? -1 : 1;
    } else if (filters.sortBy === 'amount') {
      return a.amount > b.amount ? -1 : 1;
    }
  });
};

export default getExpensesView;
