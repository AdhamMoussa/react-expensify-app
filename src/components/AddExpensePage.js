import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";

export const AddExpensePage = props => {
  const handleFormSubmit = expense => {
    props.onSubmit(expense);
    props.history.push("/");
  };
  return (
    <div>
      <div className="page-head">
        <div className="wrapper">
          <h2>Add Expense</h2>
        </div>
      </div>
      <div className="page-body">
        <div className="wrapper">
          <ExpenseForm onSubmitHandler={handleFormSubmit} />
        </div>
      </div>
    </div>
  );
};

const mapDispatchToProps = dispatch => {
  return {
    onSubmit: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
