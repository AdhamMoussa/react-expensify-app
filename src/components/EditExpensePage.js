import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export const EditExpensePage = props => {
  const onSubmitHandler = expense => {
    props.onSubmit(props.expense.id, expense);
    props.history.push("/");
  };
  const onRemoveHandler = () => {
    props.onRemoveClick(props.expense.id);
    props.history.push("/");
  };
  return (
    <div>
      <div className="page-head">
        <div className="wrapper">
          <h2>Edit Expense</h2>
        </div>
      </div>
      <div className="page-body">
        <div className="wrapper">
          <ExpenseForm
            onSubmitHandler={onSubmitHandler}
            editedExpense={props.expense}
          />
          <button
            className="button button--red"
            id="remove-btn"
            onClick={onRemoveHandler}
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  );
};

const mapDispatchToProps = dispatch => {
  return {
    onSubmit: (id, expense) => dispatch(startEditExpense(id, expense)),
    onRemoveClick: id => dispatch(startRemoveExpense(id))
  };
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(exp => exp.id === props.match.params.id)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
