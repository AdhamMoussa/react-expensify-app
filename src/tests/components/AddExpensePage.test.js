import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../fixtures/demoExpenses';

test('should render AddExpensePage correctly', () => {
  const wrapper = shallow(<AddExpensePage />);
  expect(wrapper).toMatchSnapshot();
});

test('should call onSubmit prop when onSubmitHandler called', () => {
  const onSubmitSpy = jest.fn();
  const historySpy = { push: jest.fn() };
  const wrapper = shallow(<AddExpensePage onSubmit={onSubmitSpy} history={historySpy} />);
  wrapper.find(ExpenseForm).prop('onSubmitHandler')(expenses[2]);
  expect(onSubmitSpy).toHaveBeenLastCalledWith(expenses[2]);
  expect(historySpy.push).toHaveBeenLastCalledWith('/');
});
