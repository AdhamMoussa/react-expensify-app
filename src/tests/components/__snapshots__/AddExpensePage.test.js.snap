// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <div
    className="page-head"
  >
    <div
      className="wrapper"
    >
      <h2>
        Add Expense
      </h2>
    </div>
  </div>
  <div
    className="page-body"
  >
    <div
      className="wrapper"
    >
      <ExpenseForm
        onSubmitHandler={[Function]}
      />
    </div>
  </div>
</div>
`;
